### PR TITLE
made port and bind configurable

### PR DESCRIPTION
**Pull request by behrica**
This solves #8 
my use case is to run launchpad from inside Docker.

This requires to set bind to "0.0.0.0" and have a stable port, so I can specify the port to expose while starting Docker.

### CODE REVIEW

File: src/lambdaisland/launchpad.clj [hunk 1]
```diff
@@ -380,9 +391,11 @@
                            (:extra-deps ctx)))
   ctx)
 
-(def before-steps [find-free-nrepl-port
+(def before-steps [
```
Comment (alysbrooks) [inline src/lambdaisland/launchpad.clj:394]: Seems like an extra newline?

**Review by alysbrooks — COMMENTED** [review]
Looks good. I think @plexus should review this change because I'm not sure what API he has in mind for the project as a whole—I don't want to approve this if he had other plans for `-b` and `-p`, for example.

Also, just to be clear to other reviewers, picking a port at random and choosing `127.0.0.1` were the existing behavior.

File: src/lambdaisland/launchpad.clj [hunk 2]
```diff
@@ -136,16 +141,20 @@
...
-      (assoc-dep 'refactor-nrepl/refactor-nrepl {:mvn/version (or (emacs-refactor-nrepl-version) default-refactor-nrepl-version)} ))))
+      (assoc-dep 'refactor-nrepl/refactor-nrepl {:mvn/version (or (emacs-refactor-nrepl-version) default-refactor-nrepl-version)}))))
 
-(defn find-free-nrepl-port [ctx]
-  (assoc ctx :nrepl-port (free-port)))
+(defn get-nrepl-port [ctx]
+  (assoc ctx :nrepl-port (or (get-in ctx [:options :nrepl-port])
+                          (free-port))))
+
+(defn get-nrepl-bind [ctx]
```
[the remark this answers is not in the document]
Comment (behrica) [reply src/lambdaisland/launchpad.clj:150]: I am not 100 % sure on this.
nrepl.cmd has three options related to "networking"


-b/--bind ADDR              Bind address, by default "127.0.0.1".
-h/--host ADDR              Host address to connect to when using --connect. Defaults to "127.0.0.1".
-p/--port PORT              Start nREPL on PORT. Defaults to 0 (random port) if not specified.

I think "host" is something different... Only used when using the nrepl.cmd to connect to an other repl.

We could as well allow it to be configured as part of the  "launchpad"

**Comment by behrica** [comment]
I want to add as well that nrepl itself can be configured via a "nrepl.edn" file, see here:
https://nrepl.org/nrepl/usage/server.html#server-configuration

This does not work current for launchpad, because it "overwrites" the port. So maybe this is an other way to solve the issue.

**Comment by behrica** [comment]
all three solutions are fine for my usecase (run launchpad from Docker):
- new cli options
- options in 'launchpad' file
- make sure, that the configured ":bind" and ":port" in a user supplied nrepl.edn is not overwritten

File: src/lambdaisland/launchpad.clj (hunk 1 above)
Comment (behrica) [reply src/lambdaisland/launchpad.clj:394]: yes, some emacs automation. I can remove it

**Comment by behrica** [comment]
I am now even thinking that option 3 is the best one.
It would allow to simplify the launchpad code, and remove the "find port" handling altogether.
nrepl default behavior is the same as programmed in lauchnpad (find a free port)

File: src/lambdaisland/launchpad.clj (hunk 2 above)
[the remark this answers is not in the document]
Comment (alysbrooks) [reply src/lambdaisland/launchpad.clj:150]: If it's following nREPL, then I think bind makes sense.

**Comment by alysbrooks** [comment]
@behrica I like option 3, although being able to configure it both in `nrepl.edn` and the CLI might be ideal, actually.

**Comment by plexus** [comment]
Thanks @behrica, this looks good as a first step. I have a generic mechanism in mind for setting options in `dev.local.edn`, or `~/.config/deps.edn`, with CLI args overriding `dev.local.edn` overriding `~/.clojure/deps.edn`.

As for removing the find-port handling, we need to know which port was picked later on, that's why launchpad determines the port itself.

I'm not sure yet if we want to handle `nrepl.edn`, like I said we'll have our own config system, introducing another config file to check raises questions of which one takes precedence, and sets the expectation of parity with nrepl which we may not be able to offer. (do we support every setting in `nrepl.edn`? do we need to add code every time they add a setting? what if launchpad's interpretation is not exactly the same, etc.). It also means another location to check when troubleshooting.

**Comment by plexus** [comment]
Released in

```clj
[com.lambdaisland/launchpad "0.15.79-alpha"]
{com.lambdaisland/launchpad {:mvn/version "0.15.79-alpha"}}
```

**Comment by plexus** [comment]
Released in

```clj
[com.lambdaisland/launchpad "0.15.79-alpha"]
{com.lambdaisland/launchpad {:mvn/version "0.15.79-alpha"}}
```

**Pull request merged**